grammar.js

```diff
@@ -1,157 +1,228 @@
+/* eslint-disable-next-line spaced-comment */
+/// <reference types="tree-sitter-cli/dsl" />
+// @ts-check
+
+const primitives = [
+  'blob',
+  'boolean',
+  'byte',
+  'document',
+  'double',
+  'float',
+  'integer',
+  'long',
+  'short',
+  'string',
+  'timestamp',
+  'bigInteger',
+  'bigDecimal',
+];
+
 module.exports = grammar({
   name: 'smithy',
-  extras: $ => [
+  extras: ($) => [
     /\s/,
-    ',',
+    $.comment,
     $.documentation_comment,
-    $.comment
+    ',', // commas are whitespace in smithy
   ],
 
   rules: {
     // TODO: add the actual grammar rules
-    source_file: $ =>
+    source_file: ($) =>
       seq(
         optional($.control_section),
         optional($.metadata_section),
-        optional($.shape_section)
+        optional($.shape_section),
       ),
 
-    shape_section: $ => seq(
+    control_section: ($) => repeat1($.control_statement),
+    control_statement: ($) => seq($.control_var_name, $.node_value),
+    control_var_name: ($) => seq('$', alias($.node_object_key, $.control_key), ':'),
+
+    metadata_section: ($) => repeat1($.metadata_statement),
+    metadata_statement: ($) => seq('metadata', choice($.identifier, $.string), '=', $.node_value),
+
+    shape_section: ($) => seq(
       $.namespace_statement,
-      repeat($._definition)
+      repeat($._definition),
     ),
 
-    _definition: $ => choice(
+    namespace_statement: ($) => seq('namespace', $.namespace),
+    namespace: ($) => seq($.identifier, repeat(seq('.', $.identifier))),
+
+    _definition: ($) => choice(
       $.use_statement,
-      $.apply_trait_statement,
-      $.shape_statement
+      $.apply_statement,
+      $.shape_statement,
     ),
 
-    use_statement: $ => seq('use', $.absolute_root_shape_id),
-    namespace_statement: $ => seq('namespace', $.namespace),
-    metadata_section: $ => repeat1($.metadata_statement),
-    metadata_statement: $ => seq('metadata', $.identifier, '=', $.node_value),
-    control_section: $ => repeat1($.control_statement),
-    control_statement: $ => seq($.control_var_name, $.node_value),
-    shape_statement: $ => seq(optional($.trait_statements), $.shape_body),
-    shape_body: $ => choice(
-      $.simple_shape_statement,
-      $.list_shape_statement,
-      $.map_shape_statement,
-      $.set_shape_statement,
-      $.structure_shape_statement,
-      $.union_shape_statement,
-      $.operation_shape_statement,
-      $.resource_shape_statement,
-      $.service_shape_statement,
-      $.enum_shape_statement
+    use_statement: ($) => seq('use', $.absolute_root_shape_id),
+
+    shape_statement: ($) => seq(repeat($.trait_statement), $.shape_body),
+    shape_body: ($) => choice(
+      $.simple_shape,
+      $.enum,
+      $.list,
+      $.map,
+      $.set,
+      $.structure,
+      $.union,
+      $.service,
+      $.operation,
+      $.resource,
```

Comment:
I find that a bit unnecessarily wordy, especially in the context of parsing output. Sure, for the top level items it makes sense, but imo these are all more definitions rather than statements, and it makes more sense to either suffix it with definition or leave it be (like I have)

Reply:
I think this goes against the desire you stated in your PR description:

> I tried to adhere to the IDL as much as possible

Which I agree with btw - I don't think the actual wordiness of nodes adds any overhead, but it's much clearer when comparing ABNF with the grammar.

Reply:
Yeah, this was one spot I felt very conflicted about. But I agree, maybe this should be taken up with aws then haha. I'll readd the statements in thoes names